### DISCUSSION TITLE
need help understanding "Functions" with Chat

### DISCUSSION THREAD

**Discussion by Lexian-droid** [openai-php/client#234] [Q&A]
Hello, I'm currently trying to experiment with the "Functions" system using Openai-php, however I'm struggling to figure out how this all connects together, the function isn't being called to my knowledge  Could someone advise me on how this works? I would be very appreciative.

Here is my code:

```php
require_once("./vendor/autoload.php");

$yourApiKey = "whoopsie this shouldn't be here~";
$client = OpenAI::client($yourApiKey);

function get_current_weather($location, $unit) {
    $weather_info = [
        "location" => $location,
        "temperature" => "72",
        "unit" => $unit,
        "forecast" => ["sunny", "windy"]
    ];
    return json_encode($weather_info);
}

$response = $client->chat()->create([
    'model' => 'gpt-3.5-turbo-0613',
    'messages' => [
        ['role' => 'user', 'content' => 'What\'s the weather like in New York?'],
    ],
    'functions' => [
        [
            'name' => 'get_current_weather',
            'description' => 'Get the current weather in a given location',
            'parameters' => [
                'type' => 'object',
                'properties' => [
                    'location' => [
                        'type' => 'string',
                        'description' => 'The city and state, e.g. San Francisco, CA',
                    ],
                    'unit' => [
                        'type' => 'string',
                        'enum' => ['celsius', 'fahrenheit']
                    ],
                ],
                'required' => ['location'],
            ],
        ]
    ]
]);

$response->id; // 'chatcmpl-6pMyfj1HF4QXnfvjtfzvufZSQq6Eq'
$response->object; // 'chat.completion'
$response->created; // 1677701073
$response->model; // 'gpt-3.5-turbo-0613'

foreach ($response->choices as $result) {
    $result->index; // 0
    $result->message->role; // 'assistant'
    $result->message->content; // null
    $result->message->functionCall->name; // 'get_current_weather'
    $result->message->functionCall->arguments; // "{\n  \"location\": \"Boston, MA\"\n}"
    $result->finishReason; // 'function_call'
}

$response->usage->promptTokens; // 82,
$response->usage->completionTokens; // 18,
$response->usage->totalTokens; // 100

die(json_encode($response));
```

I borrowed a bit from the documentation, as I am trying to simply figure out how it works.

**Comment by pb30** [comment]
Make sure you revoke that key, it was included in email notifications

**Comment by pb30** [comment]
It looks like it’s returning that it wants to use a function. It’s up to your code then to call your function and return the results to GPT. 

`$result->message->functionCall->name; // 'get_current_weather'
    $result->message->functionCall->arguments; // "{\n  \"location\": \"Boston, MA\"\n}"
    $result->finishReason; // 'function_call' `

**Reply by pb30** [reply]
https://platform.openai.com/docs/guides/gpt/function-calling

**Reply by vesper8** [reply]
I get this to work in Python just fine. You ask about the weather, the LLM knows this, it pipes your input to the function call, gets the result and gives it back to the LLM, and then it replies to you.

It seems like the examples provided so far only do the first step. The LLM knows it should call the function, but doesn't call it itself, so it's up to you to call it based on the initial response? But then what? How do you pass the result of the function back to the LLM so that it can then formulate a reply with the answer? How do you close the loop using openai-php/client? A complete example would be very nice.

**Comment by vesper8** [comment]
I get this to work in Python just fine. You ask about the weather, the LLM knows this, it pipes your input to the function call, gets the result and gives it back to the LLM, and then it replies to you.

It seems like the examples provided so far only do the first step. The LLM knows it should call the function, but doesn't call it itself, so it's up to you to call it based on the initial response? But then what? How do you pass the result of the function back to the LLM so that it can then formulate a reply with the answer? How do you close the loop using openai-php/client? A complete example would be very nice.

**Comment by pb30** [comment]
After you call your function, you append the results to the messages array:

```
...
['role' => 'function', 'content' => $functionResults, 'name' => 'my_func_name'],
```
